src/mongocrypt.c

```diff
@@ -574,30 +530,403 @@ mongocrypt_init (mongocrypt_t *crypt)
+   mstr_view existing_path = existing_path_.path.view;
+   bool okay = true;
+   if (!found.okay) {
+      // There is one loaded, but we failed to find that same library. Error:
+      CLIENT_ERR ("An existing CSFLE libary is loaded by the application at "
+                  "[%s], but the current call to mongocrypt_init() failed to "
+                  "find that same library.",
+                  existing_path.data);
+      okay = false;
+   } else {
+      // Get the path to what we found:
+      mcr_dll_path_result found_path = mcr_dll_path (found.lib);
+      assert (found_path.path.data && "Failed to get the DLL path of the "
+                                      "library that we just barely found");
+      if (!mstr_eq (found_path.path.view, existing_path)) {
+         // Our find-result should only ever find the existing same library.
+         // Error:
+         CLIENT_ERR (
+            "An existing CSFLE library is loaded by the application "
+            "at [%s], but the current call to mongocrypt_init() attempted to "
+            "load a second CSFLE library from [%s]. This is not allowed.",
+            existing_path.data,
+            found_path.path.data);
+         okay = false;
+      }
+      mstr_free (found_path.path);
+      mstr_free (found_path.error_string);
+   }
+
+   mstr_free (existing_path_.path);
+   mstr_free (existing_path_.error_string);
+   return okay;
+}
+
+/**
+ * @brief Drop a reference count to the global csfle loaded library.
+ *
+ * This should be called as part of mongocrypt_t destruction following a
+ * successful loading of csfle.
+ */
+static void
+_csfle_drop_global_ref ()
+{
+   mlib_call_once (&g_csfle_init_flag, init_csfle_state);
+
+   bool dropped_last_ref = false;
+   csfle_global_lib_state old_state = {.refcount = 0};
+   MONGOCRYPT_WITH_MUTEX (g_csfle_state.mtx)
+   {
+      assert (g_csfle_state.refcount > 0);
+      int new_rc = --g_csfle_state.refcount;
+      if (new_rc == 0) {
+         old_state = g_csfle_state;
+         dropped_last_ref = true;
+      }
+   }
+
+   if (dropped_last_ref) {
+      mongo_csfle_v1_status *status = old_state.vtable.status_create ();
+      const int destroy_rc =
+         old_state.vtable.lib_destroy (old_state.csfle_lib, status);
+      if (destroy_rc != MONGO_CSFLE_V1_SUCCESS && status) {
+         fprintf (stderr,
+                  "csfle lib_destroy() failed: %s [Error %d, code %d]\n",
+                  old_state.vtable.status_get_explanation (status),
+                  old_state.vtable.status_get_error (status),
+                  old_state.vtable.status_get_code (status));
+      }
+      old_state.vtable.status_destroy (status);
+
+#ifndef __linux__
+      mcr_dll_close (old_state.dll);
+#endif
+      /// NOTE: On Linux, skip closing the CSFLE library itself, since a bug in
+      /// the way ld-linux and GCC interact causes static destructors to not run
+      /// during dlclose(). Still, free the error string:
```

Comment:
As mentioned on SERVER-63710, this isn’t actually Linux-specific, and not a bug in GCC/ld-linux either, so it seems like we should a) either skip the unloading everywhere or nowhere, and b) adjust the comment:

```suggestion
      /// NOTE: Skip closing the CSFLE library itself, since the CSFLE library 
      /// performs static initialization without corresponding static cleanup,
      /// so loading it repeatedly would result in memory leaks.
```